### PR TITLE
CXP-780 Improved error message on unknown currency value

### PR DESCRIPTION
**Pull request by RiffVentura**
CXP-780 Improved error message on unknown currency value

**Description (for Contributor and Core Developer)**

Improved REST API  error message:
image

Added documentation for the error :
image

**Definition Of Done (for Core Developer only)**

- [x] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: tests/back/Pim/Enrichment/Specification/Component/Error/DocumentationBuilder/UnknownCurrencySpec.php [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    {
+        $this->beAnInstanceOf(UnknownCurrency::class);
+    }
+
+    function it_is_a_documentation_builder()
+    {
+        $this->beAnInstanceOf(DocumentationBuilderInterface::class);
+    }
+
+    function it_supports_scopable_values_constraint(ConstraintViolationInterface $constraintViolation)
```
Comment (pchasle) [inline tests/back/Pim/Enrichment/Specification/Component/Error/DocumentationBuilder/UnknownCurrencySpec.php:30]: The name of the test is not up-to-date :)

File: tests/back/Pim/Enrichment/Specification/Component/Product/Validator/Constraints/CurrencyValidatorSpec.php [hunk 2]
```diff
@@ -36,14 +37,20 @@ function it_adds_violation_when_currency_does_not_exists(
...
+        $context->getObject()->willReturn($priceCollectionValue);
+
         $price->getCurrency()->willReturn('CHF');
         $currencyRepository->getActivatedCurrencyCodes()->willReturn(['EUR', 'USD']);
-        $context->buildViolation(Argument::any())
+        $context->buildViolation(Argument::type('string'), Argument::type('array'))
             ->shouldBeCalled()
             ->willReturn($violation);
         $violation->atPath('currency')->shouldBeCalled()->willReturn($violation);
+        $violation->setCode(Argument::type('string'))->shouldBeCalled()->willReturn($violation);
```
Comment (pchasle) [inline tests/back/Pim/Enrichment/Specification/Component/Product/Validator/Constraints/CurrencyValidatorSpec.php:53]:
```suggestion
        $violation->setCode(Currency::CURRENCY)->shouldBeCalled()->willReturn($violation);
```

**Review by pchasle — COMMENTED** [review]
You might be missing the translation in validators.en_US.yml

File: src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/CurrencyValidator.php [hunk 3]
```diff
@@ -46,8 +48,12 @@ public function validate($object, Constraint $constraint)
 
         if ($object instanceof ProductPriceInterface) {
             if (!in_array($object->getCurrency(), $this->getCurrencyCodes())) {
-                $this->context->buildViolation($constraint->unitMessage)
+                $this->context->buildViolation($constraint->message, [
+                    '%attribute_code%' => $this->context->getObject()->getAttributeCode(),
```
Comment (tomglvng) [inline src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/CurrencyValidator.php:52]: function `getObject()` may return object or null. You should check before calling `getAttributeCode()`.

File: src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/Currency.php [hunk 4]
```diff
@@ -13,12 +13,9 @@
  */
 class Currency extends Constraint
 {
-    /**
-     * Violation message for invalid currency
-     *
-     * @var string
-     */
-    public $unitMessage = 'Please specify a valid currency';
+    const CURRENCY = '9b385d80-5d67-494b-b824-0b59b84d609a';
```
Comment (tseho) [inline src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/Currency.php:16]: I'm out of the loop here, why this id ?

Comment (RiffVentura) [reply src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/Currency.php:16]: The underlying validator must set a constraint code. As there isn't any , a new id is created with UUID value (repeating pattern used for other constraints)

**Review by tomglvng — APPROVED** [review]

**Review by pchasle — APPROVED** [review]

**Comment by RiffVentura** [comment]
warm master

**Pull request merged**